src/openms/include/OpenMS/ANALYSIS/ID/FalseDiscoveryRate.h

```diff
@@ -574,71 +506,97 @@ namespace OpenMS
       }
     }
 
-
     //TODO could also get a keep_decoy flag when we define what a "decoy group" is -> keep all always for now
     void setScores_(
         const std::map<double,double>& scores_to_FDR,
         std::vector<ProteinIdentification::ProteinGroup>& grps,
         const std::string& score_type,
         bool higher_better) const;
 
-    static void checkTDAnnotation_(const MetaInfoInterface & idOrHit)
-    {
-        if (!idOrHit.metaValueExists("target_decoy"))
-        {
-          throw Exception::MissingInformation(__FILE__,
-                                              __LINE__,
-                                              OPENMS_PRETTY_FUNCTION,
-                                              "Meta value 'target_decoy' does not exist in all ProteinHits! Reindex the idXML file with 'PeptideIndexer'");
-        }
-    }
+    /** @} */
 
+    /**
+     * @brief Used when keep_decoy_peptides or proteins is false
+     * @tparam HitType Protein or PeptideHit
+     * @param scores_to_FDR map from original score to FDR/qVal
+     * @param hit the Hit itself
+     * @param old_score_type to save it in metavalue
+     * @param new_hits where to move if target (i.e. target or target+decoy)
+     */
     template <typename HitType>
-    struct GetLabelFunctor: std::function<bool(const HitType&)>
+    void setScoreAndMoveIfTarget_(const std::map<double,double>& scores_to_FDR, HitType & hit, const std::string& old_score_type,
+                                  std::vector<HitType>& new_hits) const
     {
-      bool operator() (const HitType& hit)
+      const String& target_decoy(hit.getMetaValue("target_decoy"));
+      if (target_decoy[0] == 't')
       {
-        //TODO if we checked in the beginning, this check could be skipped.
-        if (!hit.metaValueExists("target_decoy"))
-        {
-          throw Exception::MissingInformation(__FILE__,
-                                              __LINE__,
-                                              OPENMS_PRETTY_FUNCTION,
-                                              "Meta value 'target_decoy' does not exist in all ProteinHits! Reindex the idXML file with 'PeptideIndexer'");
-        }
-        else
-        {
-          return std::string(hit.getMetaValue("target_decoy"))[0] == 't';
-        }
-      }
-    };
+        hit.setMetaValue(old_score_type, hit.getScore());
+        hit.setScore(scores_to_FDR.at(hit.getScore()));
```

Comment:
annotate that this throws if `hit.getScore()` is not in `scores_to_FDR`?

Reply:
To be consistent with the rest, in this case I annotate with the closest score now.